// Copyright 2009 Plant Bioinformatics Group, SCRI. All rights reserved.
// Use is subject to the accompanying licence terms.

package tablet.gui.dialog.prefs;

import java.awt.*;
import javax.swing.*;

import tablet.gui.*;

import scri.commons.gui.*;

class NBVisualizationPanel extends JPanel
{
	private DefaultComboBoxModel dnaModel;
	private DefaultComboBoxModel proteinModel;

	public NBVisualizationPanel()
    {
        initComponents();

        setBackground(Color.white);
        panel.setBackground(Color.white);

		panel.setBorder(BorderFactory.createTitledBorder(RB.getString("gui.dialog.prefs.NBVisualizationPanel.panelTitle")));

		RB.setText(dnaLabel, "gui.dialog.prefs.NBVisualizationPanel.dnaLabel");
		RB.setText(proteinLabel, "gui.dialog.prefs.NBVisualizationPanel.proteinLabel");

		dnaModel = new DefaultComboBoxModel();
        dnaModel.addElement("*");
		dnaModel.addElement("-");
        dnaCombo.setModel(dnaModel);
		dnaCombo.setSelectedIndex(Prefs.visPadCharType);

		proteinModel = new DefaultComboBoxModel();
        proteinModel.addElement(".");
		proteinModel.addElement("*");
        proteinCombo.setModel(proteinModel);
		proteinCombo.setSelectedIndex(Prefs.visStopCharType);
    }

	public void applySettings()
	{
		Prefs.visPadCharType = dnaCombo.getSelectedIndex();
		Prefs.visStopCharType = proteinCombo.getSelectedIndex();

		// This will force the visualization area to recreate its color schemes
		// which will change the rendered characters to their new states
		Tablet.winMain.getAssemblyPanel().forceRedraw();
		Tablet.winMain.getAssemblyPanel().repaint();
	}

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        panel = new javax.swing.JPanel();
        dnaLabel = new javax.swing.JLabel();
        dnaCombo = new javax.swing.JComboBox();
        proteinLabel = new javax.swing.JLabel();
        proteinCombo = new javax.swing.JComboBox();

        panel.setBorder(javax.swing.BorderFactory.createTitledBorder("Display options:"));

        dnaLabel.setLabelFor(dnaCombo);
        dnaLabel.setText("Character to display for DNA pad/gap bases:");

        proteinLabel.setLabelFor(proteinCombo);
        proteinLabel.setText("Character to display for Protein stop codons:");

        javax.swing.GroupLayout panelLayout = new javax.swing.GroupLayout(panel);
        panel.setLayout(panelLayout);
        panelLayout.setHorizontalGroup(
            panelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(panelLayout.createSequentialGroup()
                .addContainerGap()
                .addGroup(panelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(panelLayout.createSequentialGroup()
                        .addComponent(dnaLabel)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(dnaCombo, 0, 99, Short.MAX_VALUE))
                    .addGroup(panelLayout.createSequentialGroup()
                        .addComponent(proteinLabel)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(proteinCombo, 0, 99, Short.MAX_VALUE)))
                .addContainerGap())
        );
        panelLayout.setVerticalGroup(
            panelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(panelLayout.createSequentialGroup()
                .addContainerGap()
                .addGroup(panelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(dnaLabel)
                    .addComponent(dnaCombo, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(panelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(proteinLabel)
                    .addComponent(proteinCombo, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(panel, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(panel, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap(70, Short.MAX_VALUE))
        );
    }// </editor-fold>//GEN-END:initComponents

    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JComboBox dnaCombo;
    private javax.swing.JLabel dnaLabel;
    private javax.swing.JPanel panel;
    private javax.swing.JComboBox proteinCombo;
    private javax.swing.JLabel proteinLabel;
    // End of variables declaration//GEN-END:variables
}